espnet/nets/pytorch_backend/transducer/tdnn.py

```diff
@@ -1,165 +0,0 @@
-"""TDNN modules definition for custom encoder."""
-
-from typing import Tuple
-from typing import Union
-
-import torch
-
-
-class TDNN(torch.nn.Module):
-    """TDNN module with symmetric context.
-
-    Args:
-        idim: Input dimension.
-        odim: Output dimension.
-        ctx_size: Size of the context window.
-        stride: Stride of the sliding blocks.
-        dilation: Control the stride of elements within the neighborhood.
-        batch_norm: Whether to apply batch normalization.
-        relu: Whether to use a final non-linearity layer (ReLU).
-
-    """
-
-    def __init__(
-        self,
-        idim: int,
-        odim: int,
-        ctx_size: int = 5,
-        dilation: int = 1,
-        stride: int = 1,
-        batch_norm: bool = False,
-        relu: bool = True,
-        dropout_rate: float = 0.0,
-    ):
-        """Construct a TDNN object."""
-        super().__init__()
-
-        self.idim = idim
-        self.odim = odim
-
-        self.ctx_size = ctx_size
-        self.stride = stride
-        self.dilation = dilation
-
-        self.batch_norm = batch_norm
-        self.relu = relu
-
-        self.tdnn = torch.nn.Conv1d(
-            idim, odim, ctx_size, stride=stride, dilation=dilation
-        )
-
-        if self.relu:
-            self.relu_func = torch.nn.ReLU()
-
-        if self.batch_norm:
-            self.bn = torch.nn.BatchNorm1d(odim)
-
-        self.dropout = torch.nn.Dropout(p=dropout_rate)
-
-    def forward(
-        self,
-        sequence: Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]],
-        mask: torch.Tensor,
-    ) -> Tuple[Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]], torch.Tensor]:
-        """Forward TDNN for custom encoder.
-
-        Args:
-            sequence: TDNN input sequences.
-                        (B, T, D_in) or ((B, T, D_in), (B, T, D_att))
-                          or ((B, T, D_in), (B, 2 * (T - 1), D_att))
-            mask: Mask of TDNN input sequences. (B, 1, T)
-
-        Returns:
-            sequence: TDNN output sequences.
-                        (B, sub(T), D_out)
-                          or ((B, sub(T), D_out), (B, sub(T), att_dim))
-                          or ((B, T, D_out), (B, 2 * (sub(T) - 1), D_att)
-            mask: Mask of TDNN output sequences. (B, 1, sub(T))
-
-        """
-        if isinstance(sequence, tuple):
-            sequence, pos_emb = sequence[0], sequence[1]
-        else:
-            sequence, pos_emb = sequence, None
-
-        sequence = sequence.transpose(1, 2)
-
-        # The bidirect_pos is used to distinguish legacy_rel_pos and rel_pos in
-        # Conformer model. Note the `legacy_rel_pos` will be deprecated in the future.
-        # Details can be found in https://github.com/espnet/espnet/pull/2816.
-        if pos_emb is not None and pos_emb.size(1) == (2 * sequence.size(2)) - 1:
```

Comment:
@pengchengguo If I'm not mistaken, the "bidir case" is not needed? `pos_emb` will be `None` with any positional encoding class but `RelPositionalEncoding`. Also, `LegacyRelPositionalEncoding` is not supported anyways.

Reply:
Hi @b-flo,
The `bidir_pos_emb` is used to distinguish the `latest_rel_pos` and `legacy_rel_pos`. I added it because of the compatibility issues (some of the pre-trained models are trained with `legacy_rel_pos`, when w/o this param, length mismatch problem will be raised in `self.create_output()` fucntion.) For now, I think we can remove it if your models are trained with `latest_rel_pos`.

Reply:
OK, that's what I thought. Thanks for the explanation!